sway/input/keyboard.c

```diff
@@ -489,6 +489,40 @@ static void handle_key_event(struct sway_keyboard *keyboard,
 	}
 
 	if (!handled || event->state == WL_KEYBOARD_KEY_STATE_RELEASED) {
+		if (seat->im_relay.input_method) {
+			struct wlr_input_method_keyboard_grab_v2 *kb_grab =
+				seat->im_relay.input_method->keyboard_grab;
+			struct wlr_virtual_keyboard_v1 *virtual_keyboard =
+				wlr_input_device_get_virtual_keyboard(wlr_device);
+
+			// If event is from virtual keyboard of the same client as grab,
+			// do not send it back. TODO see swaywm/wlroots#2322
+			if (kb_grab && !(virtual_keyboard &&
+					wl_resource_get_client(virtual_keyboard->resource) ==
+					wl_resource_get_client(kb_grab->resource))) {
+				// Do not send release event to grab if the press event was not
+				// sent to grab.
+				if (event->state == WL_KEYBOARD_KEY_STATE_RELEASED) {
+					bool pressed_sent = update_shortcut_state(
+						&keyboard->state_pressed_sent, event->keycode,
+						event->state, keyinfo.keycode, 0);
+					if (pressed_sent) {
+						wlr_seat_set_keyboard(wlr_seat, wlr_device);
+						wlr_seat_keyboard_notify_key(wlr_seat, event->time_msec,
+							event->keycode, event->state);
+						goto end;
+					}
+
+				}
+
+				wlr_input_method_keyboard_grab_v2_set_keyboard(kb_grab,
+					wlr_device->keyboard);
+				wlr_input_method_keyboard_grab_v2_send_key(kb_grab,
+					event->time_msec, event->keycode, event->state);
+				goto end;
```

Comment:
I think we could simplify these `goto`s a little Maybe something among the lines of:

```c
	if (event->state == WL_KEYBOARD_KEY_STATE_RELEASED) {
		// If the press event was sent to a client, also send the release
		// event. In particular, don't send the release event to the IM grab.
		bool pressed_sent = update_shortcut_state(
			&keyboard->state_pressed_sent, event->keycode,
			event->state, keyinfo.keycode, 0);
		if (pressed_sent) {
			wlr_seat_set_keyboard(wlr_seat, wlr_device);
			wlr_seat_keyboard_notify_key(wlr_seat, event->time_msec,
				event->keycode, event->state);
			handled = true;
		}
	}

	if (!handled) {
		struct wlr_input_method_keyboard_grab_v2 *kb_grab =
			keyboard_get_im_grab(struct sway_keyboard *keyboard);
		if (kb_grab) {
			wlr_input_method_keyboard_grab_v2_set_keyboard(kb_grab,
				wlr_device->keyboard);
			wlr_input_method_keyboard_grab_v2_send_key(kb_grab,
				event->time_msec, event->keycode, event->state);
			handled = true;
		}
	}

	if (!handled) {
		update_shortcut_state(
			&keyboard->state_pressed_sent, event->keycode, event->state,
			keyinfo.keycode, 0);
		wlr_seat_set_keyboard(wlr_seat, wlr_device);
		wlr_seat_keyboard_notify_key(wlr_seat, event->time_msec,
			event->keycode, event->state);
	}
```

Reply:
Applied the idea. Thanks!
The code above would send released event to client even if no pressed event was sent before. I've fixed it.